### PR TITLE
Ruby Section: Advanced Building Blocks - Enumerable Methods

### PR DESCRIPTION
**Pull request by Osong-Michael**
I**n this project, I rebuilt each method in the way I understand the function or method and how they operate.**

- I also added a _test.rb_ file to help run the code in the _enum.rb_ easily and see its functionality.
- To run the code in the _test.rb_ file, make sure to uncomment the code before executing it.

### CODE REVIEW

File: enum.rb [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    self
+  end
+
+  def my_select
+    array = []
+    my_each { |index| array << index if yield(index) }
+    array
+  end
+
+  def my_all?
```
Comment (Forison) [inline enum.rb]: ### 2. Should have working `my_all?`  identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
``` p %w[ant bear cat].my_all?(/t/) ``` should return same results as ``` p %w[ant bear cat].my_all?(/t/) ```
``` p [1, 2i, 3.14].my_all?(Numeric) ```  should return same results as ``` p [1, 2i, 3.14].all?(Numeric) ```
``` p [nil, true, 99].my_all?```  should return same results as ```p[nil, true, 99].all? ```

File: enum.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    new_array = true
+    my_each { |index| break new_array = false unless yield(index) }
+    new_array
+  end
+
+  def my_any?
```
Comment (Forison) [inline enum.rb]: ### 3. Should have working `my_any?`  identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
``` p %w[ant bear cat].my_any?(/d/) ``` should return same results as ``` p %w[ant bear cat].any?(/d/) ```
``` p [nil, true, 99].my_any?(Integer)``` should return same results as ```p [nil, true, 99].any?(Integer)```
``` p [nil, true, 99].my_any? ``` shoud return same results as ```p [nil, true, 99].any?(Integer)```
``` p [12,3,34].my_any?(Integer)``` should return same results as ```p [nil, true, 99].any?(Integer)```

File: enum.rb [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    new_array = false
+    my_each { |index| break new_array = true if yield(index) }
+    new_array
+  end
+
+  def my_none?
+    new_array = true
```
Comment (Forison) [inline enum.rb]: ### 4. Should have working `my_none?`identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
``` p %w{ant bear cat}.my_none?(/d/)``` should return same results as ``` p %w{ant bear cat}.none?(/d/)```
``` p [1, 3.14, 42].my_none?(Float)``` should return same results as ```[1, 3.14, 42].none?(Float)```
``` p [nil].my_none?``` should return same results as ``` p [nil].none?```
``` p [nil, false].my_none? ``` should return same results as ``` p [nil, false].none? ``` 
``` p [nil, false, true].my_none? ``` should return same results as ``` p [nil, false, true].none? ```

File: enum.rb [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    my_each { |index| break new_array = false if yield(index) }
+    new_array
+  end
+
+  def my_count(parameter)
+    count = 0
```
Comment (Forison) [inline enum.rb:54]: ### 5. Should have working `my_count`  identical and returns the same thing as ruby's [`count`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-count). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_count.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
```ary = [1, 2, 4, 2]```
```p ary.my_count ``` should return same results as ```p ary.count ```
```p ary.my_count{ |x| x%2==0 }``` should return same results as```p ary.count{ |x| x%2==0 }```

File: enum.rb [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    my_each { |index| count += 1 if parameter == index }
+    count
+  end
+
+  def my_map(proc)
+    mapped = []
```
Comment (Forison) [inline enum.rb:69]: ### 5. Should have working `my_map` identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
``` p [1,2,3,4].map { |li| li*4 } ``` should return same results as ```p [1,2,3,4].map { |li| li*4 }```

**Review by Forison — CHANGES_REQUESTED** [review]
## Status:: Require changes🛑
# 🙋‍♂️ team,
You have done a good on this so far, here are a few issues to address to get this project approved.
### 1.```.my_each```, ```.my_each_with_index```, ```my_select```, ```.map```, should return an enumerator when no block is given.
## Check the comment 👇 ‎👇below 👇 ‎👇 and address the issues raised.
### Kindly address the issue below and submit for review
### Happy coding
Addo

**Review by stratospherique — APPROVED** [review]
### Great work implementing the suggested changes :clap: Congrats, the project is approved :heavy_check_mark: You may merge :+1: :+1: 

Keep up the good work :muscle: 
**_Ahmed Mahfoudh_**

**Pull request merged**